src/context.js

```diff
@@ -0,0 +1,191 @@
+import { ControlFunction, HaltError } from './control';
+
+export class ExecutionContext {
+  static ids = 1;
+  get isUnstarted() { return this.state === 'unstarted'; }
+  get isRunning() { return this.state === 'running'; }
+  get isWaiting() { return this.state === 'waiting'; }
+  get isCompleted() { return this.state === 'completed'; }
+  get isErrored() { return this.state === 'errored'; }
+  get isHalted() { return this.state === 'halted'; }
+
+  get isBlocking() { return this.isRunning || this.isWaiting || this.isUnstarted; }
+
+  get hasBlockingChildren() {
+    for (let child of this.children) {
+      if (child.isBlocking) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  constructor(parent = undefined, options = {}) {
+    this.id = this.constructor.ids++;
+    this.parent = parent;
+    this.options = options;
+    this.children = new Set();
+    this.exitHooks = new Set();
+    this.state = 'unstarted';
+    this.resume = this.resume.bind(this);
+    this.fail = this.fail.bind(this);
+    this.ensure = this.ensure.bind(this);
+    this.call = this.call.bind(this);
+  }
+
+  get promise() {
+    this._promise = new Promise((resolve, reject) => {
+      this.resolve = resolve;
+      this.reject = reject;
+    });
+    this.finalizePromise();
+    return this._promise;
+  }
+
+  finalizePromise() {
+    if(this.isCompleted && this.resolve) {
+      this.resolve(this.result);
+    } else if(this.isErrored && this.reject) {
+      this.reject(this.result);
+    } else if(this.isHalted && this.reject) {
+      this.reject(new HaltError(this.result));
+    }
+  }
+
+  then(...args) {
+    return this.promise.then(...args);
+  }
+
+  catch(...args) {
+    return this.promise.catch(...args);
+  }
+
+  finally(...args) {
+    return this.promise.finally(...args);
+  }
+
+  get root() {
+    if (!this.parent) {
+      return this;
+    } else {
+      return this.parent.root;
+    }
+  }
+
+  call(operation, options = {}) {
```

Comment:
I am somewhat confused by this method name, I think it doesn't describe the operation it performs very well. Maybe something like `spawn` would be better?

Reply:
The analogy that I was holding in my mind was normal function evaluation, except instead of a call stack, we have a call tree.

When you invoke a function inside most languages, the rough algorithm is:

```
1. allocate a stack frame
   1.a set arguments on the new frame
   1.b set reference to parent frame for looking up lexical variables
2. transfer control to new stack frame
3. function body runs
4. transfer control back to parent frame with return value from child frame
```

I picked `call` because that's very similar to the semantics of what we're doing here except instead of a call stack, we have a call tree.

allocate a child, passing in a reference to the parent, enter it, and then handle the result.

Reply:
I can sort of see that, but when reading and trying to understand how this fits together, the name really threw me off. I didn't understand what this was for at all, and only by reading the source did I understand that this spawns a new child, so at least the name did not fit with my mental model of what `call` would usually refer to.

And regarding the algorithm you sketched out, the last step "transfer control back to parent frame with return value from child frame" doesn't really happen when we call `call`, we get back what is essentially a Future/Promise of a value, not the value itself. Awaiting it is a separate step. And crucially the operation is executed concurrently with other operations.

I think that `spawn` is a pretty common name for this, but `fork` might also be an acceptable alternative, though it'd be nice to not overload that term.

Reply:
If I understand what you're saying, `call` has a very strong association with _synchronous_ function invocation. In fact, `call` is already a method on `Function` and the semantics of that are well known. In effection however, we could be synchronous or we could be asynchronous, and there's really no way to tell which it's going to be. On the other hand (at least for me) `spawn` and `fork` have very strong associations with _asynchronous_ execution, but again, we could be synchronous, or we might just be asynchronous. There's no way to tell.

Really, what we have is something akin to scheme's `call/cc` (call with current continuation) but that would be a silly name to use. Ideally, we can find a term that's neutral with respect to timing, but does capture the idea of doing some abstract operation within a bounded context. A litmus test is that it should read  "X this operation using that context" e.g.

* `call` this operation using that context
* `spawn` this operation using that context
* etc....

`invoke`, `start`, `begin`, `execute`, `perform`, `run` are some alternatives.

I kinda like `perform` because it is used in ember-concurrency. It also has an association with other similar efforts at structured effects:

* https://github.com/macabeus/js-proposal-algebraic-effects
* https://esdiscuss.org/topic/one-shot-delimited-continuations-with-effect-handlers

fwiw, I got the call nomenclature from https://github.com/phenax/algebraic-effects#state-effect-counter-example